cmd/ftl/cmd_deploy.go

```diff
@@ -2,266 +2,23 @@ package main
+	"github.com/TBD54566975/ftl/buildengine"
+	"github.com/TBD54566975/ftl/internal/rpc"
 )
 
 type deployCmd struct {
-	Replicas  int32  `short:"n" help:"Number of replicas to deploy." default:"1"`
-	ModuleDir string `arg:"" help:"Directory containing ftl.toml" type:"existingdir" default:"."`
-	Wait      bool   `help:"Only complete the deploy command when sufficient runners have been provisioned, i.e. the deployment is online and reachable." default:"false"`
+	Replicas int32    `short:"n" help:"Number of replicas to deploy." default:"1"`
+	Dirs     []string `arg:"" help:"Base directories containing modules." type:"existingdir" required:""`
+	NoWait   bool     `help:"Do not wait for deployment to complete." default:"false"`
 }
 
-func (d *deployCmd) Run(ctx context.Context, client ftlv1connect.ControllerServiceClient) error {
-	logger := log.FromContext(ctx)
-
-	// Load the TOML file.
-	config, err := moduleconfig.LoadModuleConfig(d.ModuleDir)
-	if err != nil {
-		return err
-	}
-
-	if len(config.Deploy) == 0 {
-		return fmt.Errorf("no deploy paths defined in config")
-	}
-
-	build := buildCmd{Dirs: []string{d.ModuleDir}}
-	err = build.Run(ctx)
-	if err != nil {
-		return err
-	}
-
-	logger.Infof("Preparing module '%s' for deployment", config.Module)
-
-	deployDir := filepath.Join(d.ModuleDir, config.DeployDir)
-	files, err := findFiles(deployDir, config.Deploy)
-	if err != nil {
-		return err
-	}
-
-	filesByHash, err := hashFiles(deployDir, files)
-	if err != nil {
-		return err
-	}
-
-	gadResp, err := client.GetArtefactDiffs(ctx, connect.NewRequest(&ftlv1.GetArtefactDiffsRequest{ClientDigests: maps.Keys(filesByHash)}))
-	if err != nil {
-		return err
-	}
-
-	module, err := d.loadProtoSchema(deployDir, config)
-	if err != nil {
-		return fmt.Errorf("failed to load protobuf schema from %q: %w", config.Schema, err)
-	}
-
-	logger.Debugf("Uploading %d/%d files", len(gadResp.Msg.MissingDigests), len(files))
-	for _, missing := range gadResp.Msg.MissingDigests {
-		file := filesByHash[missing]
-		content, err := os.ReadFile(file.localPath)
-		if err != nil {
-			return err
-		}
-		logger.Tracef("Uploading %s", relToCWD(file.localPath))
-		resp, err := client.UploadArtefact(ctx, connect.NewRequest(&ftlv1.UploadArtefactRequest{
-			Content: content,
-		}))
-		if err != nil {
-			return err
-		}
-		logger.Debugf("Uploaded %s as %s:%s", relToCWD(file.localPath), sha256.FromBytes(resp.Msg.Digest), file.Path)
-	}
-
-	resp, err := client.CreateDeployment(ctx, connect.NewRequest(&ftlv1.CreateDeploymentRequest{
-		Schema: module,
-		Artefacts: slices.Map(maps.Values(filesByHash), func(a deploymentArtefact) *ftlv1.DeploymentArtefact {
-			return a.DeploymentArtefact
-		}),
-	}))
-	if err != nil {
-		return err
-	}
-
-	_, err = client.ReplaceDeploy(ctx, connect.NewRequest(&ftlv1.ReplaceDeployRequest{DeploymentName: resp.Msg.GetDeploymentName(), MinReplicas: d.Replicas}))
+func (d *deployCmd) Run(ctx context.Context) error {
```

Comment:
FWIW both these approaches work - clients in context and clients passed directly.